pkg/validation/runner.go

```diff
@@ -0,0 +1,143 @@
+package validation
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/aws/eks-anywhere/pkg/errors"
+)
+
+// Validatable is anything that can be validated.
+type Validatable[O any] interface {
+	Copiable[O]
+}
+
+// Copiable is anything that can be copied.
+type Copiable[O any] interface {
+	DeepCopy() O
+}
+
+// Validation is the logic for a validation of a type O.
+type Validation[O Validatable[O]] func(ctx context.Context, obj O) error
+
+// Runner allows to compose and run validations.
+type Runner[O Validatable[O]] struct {
+	validations []Validation[O]
+	config      *RunnerConfig
+}
+
+// RunnerConfig contains the configuration for a Runner.
+type RunnerConfig struct {
+	maxJobs int
+}
+
+// RunnerOpt allows to configure a Runner with optional parameters.
+type RunnerOpt func(*RunnerConfig)
+
+// WithMaxJobs sets the maximun number of concurrent routines the runner will use.
+func WithMaxJobs(m int) RunnerOpt {
+	return func(c *RunnerConfig) {
+		c.maxJobs = m
+	}
+}
+
+// NewRunner constructs a new Runner.
+func NewRunner[O Validatable[O]](opts ...RunnerOpt) *Runner[O] {
+	r := &Runner[O]{
+		config: &RunnerConfig{
+			maxJobs: 8,
+		},
+	}
+
+	for _, opt := range opts {
+		opt(r.config)
+	}
+
+	return r
+}
+
+// Register adds validations to the Runner.
+func (r *Runner[O]) Register(validations ...Validation[O]) {
+	r.validations = append(r.validations, validations...)
+}
+
+// RunAll runs all validations concurrently and waits until they all finish,
+// aggregating the errors if present. obj must not be modified. If it is, this
+// indicates a programming error and the method will panic.
+func (r *Runner[O]) RunAll(ctx context.Context, obj O) errors.Aggregate {
+	copyObj := obj.DeepCopy()
+	var allErr []error
+	result := r.run(ctx, obj)
+	for err := range result {
+		allErr = append(allErr, err)
+	}
+
+	if !reflect.DeepEqual(obj, copyObj) {
+		panic("validations have modified the validated object, that is not allowed")
+	}
+
+	return errors.NewAggregate(allErr)
+}
+
+func (r *Runner[O]) run(ctx context.Context, obj O) <-chan error {
+	results := make(chan error)
+	validations := make(chan Validation[O])
+	wg := sync.WaitGroup{}
+	numWorkers := r.config.maxJobs
+	if numWorkers > len(r.validations) {
+		numWorkers = len(r.validations)
+	}
+
+	wg.Add(numWorkers)
+	for i := 0; i < numWorkers; i++ {
+		go func() {
+			for validate := range validations {
+				if err := validate(ctx, obj); err != nil {
+					for _, err := range flatten(err) {
```

Comment:
I think we can leave flattening to `RunAll`.

Reply:
Fair. I only did it this way because of the possibility of having a `Run` as mentioned in https://github.com/aws/eks-anywhere/pull/4672#discussion_r1110151211
I'll update this accordingly to the resolution of that comment